packages/material-ui/src/OverridableComponent.d.ts

```diff
@@ -0,0 +1,50 @@
+import * as React from 'react';
+import { ConsistentWith, PropsOf, Omit } from '.';
+import { StyledComponentProps } from './styles';
+
+export interface OverridableComponent<M extends OverridableTypeMap> {
+  <C extends AnyReactType>(props: { component: C } & OverrideProps<M, C>): JSX.Element;
+  (props: DefaultProps<M>): JSX.Element;
+}
```

Comment:
Note that this means we are declaring all overridable components to be function components, which is almost certainly a lie currently. But given the direction React is headed with Hooks, it will mostly likely be true someday? In any case, this was the only way I could get this approach to work.

Reply:
That will mean however that passing a `ref` is forbidden with the typings. Since we want `withStyles` to return `forwardRef` components in v4 this is quite an issue. This would already be an issue for v3 because every component is a class component (every component is decorated with `withStyles`) and can hold refs. The new typings would not allow that.

Reply:
I tried this:

```ts
export interface OverridableComponent<M extends OverridableTypeMap> {
  new <C extends AnyReactType>(props: { component: C } & OverrideProps<M, C>): React.Component<{ component: C } & OverrideProps<M, C>>;
  new (props: DefaultProps<M>): React.Component<DefaultProps<M>>;
}
```

but it seems to then allow any props at all when overriding the root component, e.g.

```ts
// no type error:
<Button component={(props: {}) => null} bogusProps={42} />;
```

I'll file a TS bug report about this when I have some time.

Reply:
Hi! I saw this PR linked from the cheatsheet repo. Have you tried just intersecting the `props` parameter to have `ref`? Like what @eps1lon said, it shouldn't be any "trick" TypeScript itself doesn't actually treat `ref` specially, so it will happily accept `ref` as an acceptable prop for a function component. What it actually does is that it always injects `ref` for class and intrinsic components, but you can add it manually for all other cases.

Reply:
I just tried using refs with no additional changes to the PR and it seems to already work, and has the appropriate `forwardRef` behavior?

```ts
<Button
  ref={elem => {
    elem; // $ExpectType HTMLButtonElement | null
  }}
/>
<Button<'a'>
  component="a"
  ref={elem => {
    elem; // $ExpectType HTMLAnchorElement | null
  }}
/>
```

Reply:
@ferdaber Thank you for clarifying that. I guess it makes sense to treat every prop equal at a type level. I just hope there isn't any lint rule that sees a function and automatically assumes you can't attach `ref`.

Reply:
@pelotom Awesome! I need to go over the PR again to familiarize myself with the approach. Maybe add a test for the ref just to be sure. Especially what happens if you use other class or function components in the `component` prop.

Reply:
Hm, although it seems not to handle refs for class components currently... I'll have to look into that later.

Reply:
The `ref` from a class component is injected by TypeScript inside a JSX expression, so you'll probably need to use `ComponentPropsWithRef` in `OverrideProps`: something like `OverrideProps<M, ComponentPropsWithRef<C>>` (though I think `OverrideProps` already tries to extract the props out of `C`, but that's the gist of it).